src/Integration/RuleSets/SonarQubeRuleSetProvider.cs

```diff
@@ -0,0 +1,169 @@
+    {
+        private readonly ISonarQubeService sonarQubeService;
+        private readonly ILogger logger;
+
+        public SonarQubeRuleSetProvider(ISonarQubeService sonarQubeService, ILogger logger)
+        {
+            if (sonarQubeService == null)
+            {
+                throw new ArgumentNullException(nameof(sonarQubeService));
+            }
+
+            if (logger == null)
+            {
+                throw new ArgumentNullException(nameof(logger));
+            }
+
+            this.sonarQubeService = sonarQubeService;
+            this.logger = logger;
+        }
+
+        public SonarRuleSet GetRuleSet(BoundSonarQubeProject project, Language language)
+        {
+            if (project == null)
+            {
+                throw new ArgumentNullException(nameof(project));
+            }
+
+            if (language == null)
+            {
+                throw new ArgumentNullException(nameof(language));
+            }
+
+            if (language != Language.CSharp && language != Language.VBNET)
+            {
+                throw new ArgumentOutOfRangeException(nameof(language));
+            }
+
+            var roslynProfileExporter = GetSonarQubeRuleSet(project, language).GetAwaiter().GetResult();
+            if (roslynProfileExporter == null)
+            {
+                return null;
+            }
+
+            var ruleset = ConvertToCodeAnalysisRuleSet(roslynProfileExporter);
+
+            return ConvertToSonarRuleSet(ruleset, language);
+        }
+
+        private async Task<RoslynExportProfileResponse> GetSonarQubeRuleSet(BoundSonarQubeProject project, Language language)
+        {
+            var serverLanguage = language.ToServerLanguage();
+
+            var qualityProfileInfo = await SafeServiceCall(() => this.sonarQubeService.GetQualityProfileAsync(
+                project.ProjectKey, project.Organization?.Key, serverLanguage, CancellationToken.None));
+            if (qualityProfileInfo == null)
+            {
+                this.logger.WriteLine(string.Format(Strings.SubTextPaddingFormat,
+                   string.Format(Strings.CannotDownloadQualityProfileForLanguage, language.Name)));
+                return null;
+            }
+
+            var roslynProfileExporter = await SafeServiceCall(() => this.sonarQubeService.GetRoslynExportProfileAsync(
+                qualityProfileInfo.Name, project.Organization?.Key, serverLanguage, CancellationToken.None));
+            if (roslynProfileExporter == null)
+            {
+                this.logger.WriteLine(string.Format(Strings.SubTextPaddingFormat,
+                    string.Format(Strings.QualityProfileDownloadFailedMessageFormat, qualityProfileInfo.Name,
+                        qualityProfileInfo.Key, language.Name)));
+                return null;
+            }
+
+            return roslynProfileExporter;
+        }
+
+        private async Task<T> SafeServiceCall<T>(Func<Task<T>> call)
```

Comment:
This helper method is always a copy-paste from `BindingWorkflow`.

Reply:
Maybe we should move it to shared class, or even add _SafeXXX_ extensions to the service methods.